common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,487 @@
+    public String getString(PartStyle.StyleType type) {
+        if (type == PartStyle.StyleType.FULL) {
+            return ComponentUtils.getCoded(temporaryWorkaround);
+        }
+
+        StringBuilder builder = new StringBuilder();
+
+        PartStyle previousStyle = null;
+        for (StyledTextPart part : parts) {
+            builder.append(part.getString(previousStyle, type));
+            previousStyle = part.getPartStyle();
+        }
+
+        return builder.toString();
+    }
+
+    public MutableComponent getComponent() {
+        if (parts.isEmpty()) {
+            return Component.empty();
+        }
+
+        MutableComponent component = parts.get(0).getComponent();
+
+        for (int i = 1; i < parts.size(); i++) {
+            component.append(parts.get(i).getComponent());
+        }
+
+        return component;
+    }
+
+    public static StyledText join(String separator, StyledText... texts) {
+        List<StyledTextPart> parts = new LinkedList<>();
+        List<ClickEvent> clickEvents = new LinkedList<>();
+        List<HoverEvent> hoverEvents = new LinkedList<>();
+
+        final int length = texts.length;
+        for (int i = 0; i < length; i++) {
+            StyledText text = texts[i];
+            parts.addAll(text.parts);
+
+            if (i != length - 1) {
+                // Possible micro-optimization: If the separator does not have formatting, use the normal constructor
+                parts.add(StyledTextPart.fromComponentWithPossibleFormattingCodes(separator, Style.EMPTY, null, null));
+            }
+
+            clickEvents.addAll(text.clickEvents);
+            hoverEvents.addAll(text.hoverEvents);
+        }
+
+        return new StyledText(parts, clickEvents, hoverEvents);
+    }
+
+    public static StyledText join(String separator, Iterable<StyledText> texts) {
+        return join(separator, Iterables.toArray(texts, StyledText.class));
+    }
+
+    public static StyledText concat(StyledText... texts) {
+        List<StyledTextPart> parts = new LinkedList<>();
+        List<ClickEvent> clickEvents = new LinkedList<>();
+        List<HoverEvent> hoverEvents = new LinkedList<>();
+
+        for (StyledText text : texts) {
+            parts.addAll(text.parts);
+            clickEvents.addAll(text.clickEvents);
+            hoverEvents.addAll(text.hoverEvents);
+        }
+
+        return new StyledText(parts, clickEvents, hoverEvents);
+    }
+
+    public static StyledText concat(Iterable<StyledText> texts) {
+        return concat(Iterables.toArray(texts, StyledText.class));
+    }
+
+    public StyledText getNormalized() {
+        return new StyledText(
+                parts.stream().map(StyledTextPart::asNormalized).collect(Collectors.toList()),
+                temporaryWorkaround,
+                clickEvents,
+                hoverEvents);
+    }
+
+    public StyledText trim() {
+        return new StyledText(
```

Comment:
This is not correct. trim means removing trailing and leading whitespace for the entire text, not for each part.